### DISCUSSION TITLE
How to create cure flags? Cure flag is a common field to create in credit risk task but I can't figure out how to create it

### DISCUSSION THREAD

**Discussion by xiaodaigh** [duckdb/duckdb#14137] [Q&A]
In credit risk modelling, the data is often organised in monthly snapshots. We often create a binary flag called `default_12m` which is `TRUE` if the customer's delinquency `delq` goes above or equal to 3. Delinquency can be understood as number of monthly bills not paid by the customer. So `delq=2` means the customer hasn't paid the bills for 2 months.

So I had no problem creating the `default_12m` flag. See below

```sql
library("duckdb")
con <- dbConnect(duckdb(), dbdir = ":memory:")


a = data.frame(
  date=seq(as.Date("2018-11-1"), as.Date("2024-09-01"), by = "month"),
  delq = c(rep(0, 11), 1:3, rep(0, 57))
  )

dbWriteTable(con, "testing", a)

res <- dbExecute(con, 
"
  create or replace table abc as 
  SELECT 
    date,
    delq,
    sum(case when delq>=3 then 1 else 0 end) over (order by date ROWS Between 1 following and 12 following) as default_12m
  FROM
    testing
")
print(res)

res <- dbGetQuery(con,                   "
  SELECT 
    *
  FROM
    abc
")
print(res)
```
which shows this
![image](https://github.com/user-attachments/assets/27bec29b-6415-400d-8842-e3daa368af5f)

But I can't for the life of me figure out how to create the cure flag after going through the duckdb docs on list functions, windows, and windowing functions. I know this is more of a SQL question than duckdb question but I just wanted to see if someone can show me how to do it in duckdb SQL. 

So here's how the cure flag works: once the customer has defaulted i.e. `delq=3`. The customer can start paying off the bills so his/her `delq` can go down. Once `delq` reaches 0, the customer is said to have entered `curing` status. If the customer's `delq` can remain at 0 for 12 months, then the status changes to `cured`. However, if during curing, the customer's `delq` goes above 0, then the curing counter restarts and he has to reach 0 and stay there for 12 months again to be `cured`. Because the curing/cured status depends on if the customer has entered `delq=3` in the past and I don't even know how to start.

This is relatively easy to write in a procedure language but in a declarative one I can't figure it out. It's a bit late so I will leave this question here and be back tomorrow to provide what the output should look like. But appreciate the help if anyone can give me tips on what to do.

**Comment by soerenwolfers** [comment]
Great use case for https://github.com/duckdb/duckdb/discussions/10703 

 In absence of that, you can look at the workarounds I gave there, which should be easy to adapt to your problem. Let me know if you can't work it out yourself. I'm fairly sure there is no easy solution: you will always have to either use recursive CTEs or inefficient and awkward `list_reduce`s. 

EDIT: Also, great use case for https://github.com/duckdb/duckdb/discussions/3994

**Reply by xiaodaigh** [reply]
I think reduce is not easy but maybe we need an accumulator. Or a transform that also gives access to previous elements. This might be one of the reasons why SAS is used. To create flags like this. In the real work setting there are even more logic to take care of so it’s hard in a declarative language like sql

**Reply by xiaodaigh** [reply]
I got an answer for a related problem using cte

**Reply by soerenwolfers** [reply]
> I think reduce is not easy but maybe we need an accumulator


I don't know what you mean by this.

**Reply by xiaodaigh** [reply]
Reduce reduces to one numbers. I want to return a whole vector

**Reply by soerenwolfers** [reply]
did you look at the list_reduce solution I gave at #10703? It does produce one output row per input row. It does so by applying the per-row function `list_reduce` to the result of a window function that collects all previous rows into a list, for every input row.

**Reply by xiaodaigh** [reply]
Will check it out
